torch/csrc/jit/pickler.cpp

```diff
@@ -1,54 +1,81 @@
 #include <torch/csrc/jit/pickler.h>
+#include <string>
 
 namespace torch {
 namespace jit {
 
 using ::c10::IValue;
 
-PicklerClass getClass(const std::string& str) {
-  if (str == "TensorID") {
-    return PicklerClass::TENSOR;
-  } else if (str == "IntList") {
-    return PicklerClass::INTLIST;
-  }
-  AT_ERROR("Unknown class name for unpickler: ", str);
-}
+const static std::unordered_map<std::string, PicklerClass> name_to_class{
+  {"TensorID", PicklerClass::TENSOR},
+  {"IntList", PicklerClass::INTLIST},
+};
 
-const std::string& getClassName(PicklerClass cls) {
-  static const std::string tensor_class("TensorID\n");
-  static const std::string intlist_class("IntList\n");
-  switch (cls) {
-    case PicklerClass::TENSOR:
-      return tensor_class;
-    case PicklerClass::INTLIST:
-      return intlist_class;
-    default:
-      AT_ERROR("Unknown class for pickler");
-  }
-}
+const static std::unordered_map<PicklerClass, std::string, std::hash<uint8_t>>
+    class_to_name{
+        {PicklerClass::TENSOR, "TensorID\n"},
+        {PicklerClass::INTLIST, "IntList\n"},
+    };
 
-const std::string& getModuleName() {
-  static const std::string module_name("__main__\n");
-  return module_name;
-}
+static std::string module_name = "__main__\n";
+
+// Protocol 2 is the highest that can be decoded by Python 2
+// See https://docs.python.org/3/library/pickle.html#data-stream-format
+constexpr static uint8_t PROTOCOL_VERSION = 2;
 
 const std::vector<char>& Pickler::stack() {
   return stack_;
 }
 
-void Pickler::start() {
+void Pickler::start(bool wrap_in_list) {
   push<OpCode>(OpCode::PROTO);
-  push<uint8_t>(2);
+  push<uint8_t>(PROTOCOL_VERSION);
 
   // All attributes get pushed into a list and their indices saved in the
   // module def
-  push<OpCode>(OpCode::EMPTY_LIST);
-  push<OpCode>(OpCode::MARK);
+  wrap_in_list_ = wrap_in_list;
+  if (wrap_in_list_) {
+    push<OpCode>(OpCode::EMPTY_LIST);
+    push<OpCode>(OpCode::MARK);
+  }
 }
 
 void Pickler::finish() {
-  push<OpCode>(OpCode::APPENDS);
+  if (wrap_in_list_) {
+    push<OpCode>(OpCode::APPENDS);
+  }
   push<OpCode>(OpCode::STOP);
+
+  // Add the binary data for all the tensors to be included in the same binary
+  // TODO: The pickler should be refactored to stream out to a file directly
+  // instead of staging in the stack_ array
+
+  // Insert storage keys
+  if (literal_tensors_.size() > 0) {
```

Comment:
This organization is weird. I have to do a lot of self-convincing that this isn't an infinite recursion. Can this be refactored to make it more clear?